src/sycltest/SYCLCore/device_unique_ptr.h

```diff
@@ -0,0 +1,96 @@
+#ifndef HeterogeneousCore_SYCLUtilities_interface_device_unique_ptr_h
+#define HeterogeneousCore_SYCLUtilities_interface_device_unique_ptr_h
+
+#include <functional>
+#include <memory>
+#include <optional>
+
+#include <CL/sycl.hpp>
+
+namespace cms {
+  namespace sycltools {
+    namespace device {
+      namespace impl {
+        // Additional layer of types to distinguish from host::unique_ptr
+        class DeviceDeleter {
+        public:
+          DeviceDeleter() = default;  // for edm::Wrapper
+          DeviceDeleter(sycl::queue stream) : stream_{stream} {}
+
+          void operator()(void *ptr) {
+            if (stream_) {
+              sycl::free(ptr, *stream_);
+            }
+          }
+
+        private:
+          std::optional<sycl::queue> stream_;
+        };
+      }  // namespace impl
+
+      template <typename T>
+      using unique_ptr = std::unique_ptr<T, impl::DeviceDeleter>;
+
+      namespace impl {
+        template <typename T>
+        struct make_device_unique_selector {
+          using non_array = cms::sycltools::device::unique_ptr<T>;
+        };
+        template <typename T>
+        struct make_device_unique_selector<T[]> {
+          using unbounded_array = cms::sycltools::device::unique_ptr<T[]>;
+        };
+        template <typename T, size_t N>
+        struct make_device_unique_selector<T[N]> {
+          struct bounded_array {};
+        };
+      }  // namespace impl
+    }    // namespace device
+
+    template <typename T>
+    typename device::impl::make_device_unique_selector<T>::non_array make_device_unique(sycl::queue stream) {
+      static_assert(std::is_trivially_constructible<T>::value,
+                    "Allocating with non-trivial constructor on the device memory is not supported");
+      void *mem = sycl::malloc_device(sizeof(T), stream);
+      return typename device::impl::make_device_unique_selector<T>::non_array{reinterpret_cast<T *>(mem),
+                                                                              device::impl::DeviceDeleter{stream}};
+    }
+
+    template <typename T>
+    typename device::impl::make_device_unique_selector<T>::unbounded_array make_device_unique(size_t n,
+                                                                                              sycl::queue stream) {
+      using element_type = typename std::remove_extent<T>::type;
+      static_assert(std::is_trivially_constructible<element_type>::value,
+                    "Allocating with non-trivial constructor on the device memory is not supported");
+      void *mem = sycl::malloc_device(n * sizeof(element_type), stream);
```

Comment:
Do you know (or can you point to documentation) how the `malloc_device()` (and `malloc_host()`) use the `sycl::queue`?

Reply:
Documentation is scarce...

According to the [SYCL 2020 specification](https://www.khronos.org/registry/SYCL/specs/sycl-2020-provisional.pdf#page=200), the `sycl::queue` is used just to get the `sycl::device` and `sycl::context` to be used for the memory allocation.

That is, there is no mention of any synchronisation with the host or device work in the queue.

Reply:
Ok, thanks.